torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,803 @@
+        GPUs. This shard will align with the current GPU rank. For example::
+            before reduce_scatter:
+                param.grad (GPU #0): [1, 2, 3, 4]
+                param.grad (GPU #1): [5, 6, 7, 8]
+            after reduce_scatter:
+                param.grad (GPU #0): [6, 8]    # 1+5, 2+6
+                param.grad (GPU #1): [10, 12]  # 3+7, 4+8
+        The local GPU's ``optim.step`` is responsible for updating a single
+        shard of params, also corresponding to the current GPU's rank. This
+        alignment is created by :func:`_shard_parameters_`, which ensures that
+        the local optimizer only sees the relevant parameter shard.
+        """
+        # First hook callback will see PRE state. If we have multiple params,
+        # then subsequent hook callbacks will see POST state.
+        self._assert_state([TrainingState.BACKWARD_PRE, TrainingState.BACKWARD_POST])
+        self.training_state = TrainingState.BACKWARD_POST
+        if param.grad is None:
+            return
+
+        if param.grad.requires_grad:
+            raise RuntimeError("FSDP only works with gradients that don't require gradients")
+
+        self._free_full_params([param])
+        # Switch to local shard after backward.
+        self._use_param_local_shard([param])
+
+        # Wait for all work in the current stream to finish, then start the
+        # reductions in post_backward stream.
+        self._streams["post_backward"].wait_stream(torch.cuda.current_stream())
+        with torch.cuda.stream(self._streams["post_backward"]):
+            orig_grad_data = param.grad.data
+
+            if self.gradient_predivide_factor > 1:
+                # Average grad by world_size for consistency with PyTorch DDP.
+                param.grad.div_(self.gradient_predivide_factor)
+
+            def _post_reduction_hook(obj) -> None:
+                """Hook to call on each param after the reduce-scatter."""
+                # If obj is a tensor, current stream is post_backward stream;
+                # If obj is a future, current stream is a new temporary team,
+                # use post_backward stream explicitly here, so that later on
+                # computation stream can wait for operations here
+                self._streams["post_backward"].wait_stream(torch.cuda.current_stream())
+                with torch.cuda.stream(self._streams["post_backward"]):
+                    assert param.grad is not None
+                    self._assert_state(TrainingState.BACKWARD_POST)
+                    if type(obj) == torch.Tensor:
+                        reduced_grad = obj
+                    else:
+                        reduced_grad = obj.value()[0]
+                    param.grad.data = reduced_grad
+                    if self.gradient_postdivide_factor > 1:
+                        # Average grad by world_size for consistency with PyTorch DDP.
+                        param.grad.div_(self.gradient_postdivide_factor)
```

Comment:
(no need to change this PR) do we need a separate stream for the computation, so that it's not blocking subsequent communication ops?

Reply:
sounds good